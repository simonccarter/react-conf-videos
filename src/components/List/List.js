import * as React from 'react'
import { connect } from 'react-redux'
import { flatten, map, pathOr } from 'ramda'
import { compose, pure, withHandlers } from 'recompose'

import { Conference } from '../../domain'
import { SearchInput } from '../SearchInput'
import { Video } from '../Videos/Video.tsx'

import styles from './List.scss'

// type Props = {
//   filter: any
//   setIsActive: any
// }

// type WithHandlers = {
//   onInputChange: (e: any) => void
// }

// type State = {
//   conferences: {[idx: string]: Conference}
// }

// type CombinedProps = Props & WithHandlers & State

const ListInner = ({ conferences }) => {
  const children = flatten(Object.keys(conferences).map((conferenceId) => {
    const conferenceProps = conferences[conferenceId]
    return map(video => <Video key={video} videoId={video} conferenceId={conferenceId} />, pathOr([], ['videos'], conferenceProps))
  }))
  const countVids = children.length
  const countVidsS = countVids === 1 ? 'video' : 'videos'
  const countConfs = Object.keys(conferences).length
  const countConfsS = countConfs === 1 ? 'conference' : 'conferences'

  return  (
    <div className={styles.root}>
      <SearchInput />
      <p className={styles.resultsCount}>
        <span className={styles.resultsNumber}> {countVids} </span> {countVidsS}
        <span className={styles.resultsNumber}> {countConfs} </span> {countConfsS}
      </p>
      <section className={styles.results}>
        { true && children }
      </section>
    </div>
  )
}

const mapStateToProps = ({ frontPage }) => ({
  conferences: frontPage.filteredConferences,
})

const List = compose(
  connect(mapStateToProps),
  pure,
  withHandlers({
    onInputChange: ({ filter, setIsActive }) => (e) => {
      setIsActive(e.target.value !== '')
      filter(e.target.value)
    }
  })
)(ListInner)

export default List
