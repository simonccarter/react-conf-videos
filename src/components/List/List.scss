.root {
  padding-bottom: 2rem;
}