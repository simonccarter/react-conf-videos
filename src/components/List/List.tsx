import * as React from 'react'
import { mapProps } from 'recompose'
import { compose as rcompose, flatten, map, pathOr, toPairs } from 'ramda'

import { Video } from 'components'
import { Conference, IndexedConferences } from '../../domain'

import styles from './List.scss'

type Props = { conferences: {[idx: string]: Conference} }
type MapProps = { videos: any[] }

const mapConferenceIdOntoVideos = ([conferenceId, conference]: [string, Conference]) =>
  map(
    (video => (<Video key={video} videoId={video} conferenceId={conferenceId} />)), 
    pathOr([], ['videos'], conference)
  )

const mapConferenceVideos = rcompose<IndexedConferences, any, any, any[]>(
  flatten,
  map(mapConferenceIdOntoVideos),
  toPairs
)

export const ListInner: React.SFC<MapProps> = ({ videos }) => {
  return (
    <section className={styles.root}> 
      { videos }
    </section>
  )
}

const List = mapProps<MapProps, Props>(({conferences}) => 
  ({ videos: mapConferenceVideos(conferences)})
)(ListInner)

export default List
