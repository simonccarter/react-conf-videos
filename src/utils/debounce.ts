export default (func: (...idx: unknown[]) => void, delay = 300) => {
  let timer: NodeJS.Timeout;

  return (...args: unknown[]) => {
    clearTimeout(timer);
    timer = setTimeout(() => func(...args), delay);
  };
};
