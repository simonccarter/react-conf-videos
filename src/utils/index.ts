export * from './test'
export * from './search'
export * from './url'