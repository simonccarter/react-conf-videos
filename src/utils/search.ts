const isFilterEmpty = (filterValue: string = '') => filterValue.trim() === ''

export { isFilterEmpty }