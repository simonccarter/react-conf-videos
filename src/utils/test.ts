import { createMockStore } from 'redux-test-utils'
import { MiddlewareAPI } from 'redux'
import { mount, shallow } from 'enzyme'
import { ReactElement } from 'react'

import { Conference } from '../domain'

/** epics **/

// fails test of error cb of subcribe is called
export const onError = (done: any) => (error: Error) => done(false)

// mock empty store object
export const mockStore = () => { return <MiddlewareAPI<void>>{} }

/** Components **/

const isStore = (store: any) => store.hasOwnProperty('isActionTypeDispatched')

// shallow mount component with a store in context
export const shallowWithStore = <T>(storeP: T, component: ReactElement<any>) => {
  const store = isStore(storeP) ? storeP : createMockStore(storeP)
  return shallow<any, any>(component, { context: { store } })
}

export const mountWithStore = <T>(storeP: T, component: ReactElement<any>) => {
  const store = isStore(storeP) ? storeP : createMockStore(storeP)
  return mount<any, any>(component, { context: { store } })
}

/** Mock data */
export const mockConference = (): Conference => ({
  date: 'XX/YY/ZZZ',
  title: 'fake conf title',
  website: 'fake url',
  playlist: 'day 1',
  videos: ['aaa', 'bbb', 'ccc']
})