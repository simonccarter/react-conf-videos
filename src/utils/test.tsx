import * as React from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { RecoilRoot } from 'recoil';
import ScrollToTop from 'components/HOCS/ScrollToTop';
import {
  ConferenceTransformed,
  VideoTransformed,
} from '../domain/TransformedJSON';

export const mockVideo = (title = 'test title'): VideoTransformed => ({
  id: 'SOME_RANDOM_VIDEO_ID',
  link: 'a link',
  split: '',
  title: title || 'test title',
  length: '12:34',
  lightning: false,
  presenter: 'simon carter',
  embeddableLink: 'a link',
  conference: {
    date: 'XX/YY/ZZZ',
    title: 'react conf 2018',
    website: 'fake url',
    playlist: 'day 1',
  },
});

export const mockConference = (): ConferenceTransformed => ({
  date: 'XX/YY/ZZZ',
  title: 'react conf 2018',
  website: 'fake url',
  playlist: 'day 1',
  videos: [mockVideo('aaa'), mockVideo('bbb'), mockVideo('ccc')],
});

export const Providers: React.FC = ({ children }) => (
  <Router>
    <RecoilRoot>
      <ScrollToTop>{children}</ScrollToTop>
    </RecoilRoot>
  </Router>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>
) => render(ui, { wrapper: Providers, ...options });

export * from '@testing-library/react';
export { customRender as render };
