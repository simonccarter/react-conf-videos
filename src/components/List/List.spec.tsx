import * as React from 'react'
import { shallow } from 'enzyme'
import toJSON from 'enzyme-to-json'

import { mockConference } from 'utils'

import List from './List'

describe('List', () => {
  describe('ListInner', () => {
    it('should render', () => {
      // arrange
      const conf = mockConference()
      const props = { conferences: {'yyy': conf}, displayConferenceDetails: true }

      // act
      const wrapper = shallow(<List {...props} />)

      // assert
      expect(toJSON(wrapper)).toMatchSnapshot()
    })
  })
})