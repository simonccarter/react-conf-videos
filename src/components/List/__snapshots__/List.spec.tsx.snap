// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List ListInner should render 1`] = `
<Component
  videos={
    Array [
      <Connect(pure(withStateHandlers(withHandlers(Component))))
        conferenceId="yyy"
        videoId="aaa"
      />,
      <Connect(pure(withStateHandlers(withHandlers(Component))))
        conferenceId="yyy"
        videoId="bbb"
      />,
      <Connect(pure(withStateHandlers(withHandlers(Component))))
        conferenceId="yyy"
        videoId="ccc"
      />,
    ]
  }
/>
`;
